Source/VMware.vSphereDSC/DSCResources/Cluster/HACluster.ps1

```diff
@@ -0,0 +1,276 @@
+
+    <#
+    .DESCRIPTION
+
+    Specifies a configured failover level.
+    This is the number of physical host failures that can be tolerated without impacting the ability to meet minimum thresholds for all running virtual machines.
+    The valid values range from 1 to 4.
+    #>
+    [DscProperty()]
+    [nullable[int]] $HAFailoverLevel
+
+    <#
+    .DESCRIPTION
+
+    Indicates that the virtual machine should be powered off if a host determines that it is isolated from the rest of the compute resource.
+    The valid values are PowerOff, DoNothing and Unset.
+    #>
+    [DscProperty()]
+    [HAIsolationResponse] $HAIsolationResponse = [HAIsolationResponse]::Unset
+
+    <#
+    .DESCRIPTION
+
+    Specifies the cluster HA restart priority. The valid values are Unset, Disabled, Low, Medium, and High.
+    VMware HA is a feature that detects failed virtual machines and automatically restarts them on alternative ESX hosts.
+    #>
+    [DscProperty()]
+    [HARestartPriority] $HARestartPriority = [HARestartPriority]::Unset
+
+    hidden [string] $HAEnabledParameterName = 'HAEnabled'
+    hidden [string] $HAAdmissionControlEnabledParameterName = 'HAAdmissionControlEnabled'
+    hidden [string] $HAFailoverLevelParameterName = 'HAFailoverLevel'
+    hidden [string] $HAIsolationResponseParameterName = 'HAIsolationResponse'
+    hidden [string] $HARestartPriorityParemeterName = 'HARestartPriority'
+
+    [void] Set() {
+        $clusterData = $this.GetClusterData()
+
+        $clusterLocation = $clusterData.ClusterLocation
+        $cluster = $clusterData.Cluster
+
+        if ($this.Ensure -eq [Ensure]::Present) {
+            if ($null -eq $cluster) {
+                $this.AddCluster($clusterLocation)
+            }
+            else {
+                $this.UpdateCluster($cluster)
+            }
+        }
+        else {
+            if ($null -ne $cluster) {
+                $this.RemoveCluster($cluster)
+            }
+        }
+    }
+
+    [bool] Test() {
+        $clusterData = $this.GetClusterData()
+        $cluster = $clusterData.Cluster
+
+        if ($this.Ensure -eq [Ensure]::Present) {
+            if ($null -eq $cluster) {
+                return $false
+            }
+
+            return !$this.ShouldUpdateCluster($cluster)
+        }
+        else {
+            return ($null -eq $cluster)
+        }
+    }
+
+    [HACluster] Get() {
+        $result = [HACluster]::new()
+        $result.Server = $this.Server
+        $result.InventoryPath = $this.InventoryPath
+        $result.Datacenter = $this.Datacenter
+
+        $clusterData = $this.GetClusterData()
+        $cluster = $clusterData.Cluster
+
+        $this.PopulateResult($cluster, $result)
+
+        return $result
+    }
+
+    <#
+    .DESCRIPTION
+
+    Returns a hashtable containing the Cluster Location and the Cluster with the specified name if if exists,
+    otherwise the Cluster is set to $null.
+    #>
+    [hashtable] GetClusterData() {
```

Comment:
ClusterData is a bit confusing for the object it returns, I'd suggest ClusterInventoryItem with properties:

'Location'
'Cluster'. 

I'm wondering why is this needed at all, Cluster object has the ParentFolder folder property, why not just GetCluster() if could be null or valid with ParentFolder